src/lib.rs

```diff
@@ -511,3 +555,315 @@ where
+}
+
+impl ValueBuffer {
+    fn new() -> Self {
+        Self {
+            last_offset: None,
+            values: Vec::new(),
+        }
+    }
+
+    fn add(&mut self, value: Value, offset: DataTypeOffset) {
+        self.last_offset = Some(offset);
+        self.values.push(value);
+    }
+
+    fn reset(&mut self, offset: Option<DataTypeOffset>) {
+        self.last_offset = offset;
+        self.values.clear();
+    }
+
+    fn consume(&mut self) -> (Vec<Value>, DataTypeOffset) {
+        let consumed = (self.values.drain(0..).collect(), self.last_offset.unwrap());
+        self.last_offset = None;
+        consumed
+    }
+
+    fn non_empty(&self) -> bool {
+        self.last_offset != None
+    }
+}
+
+pub struct Context {
+    partition_assignment: Arc<Mutex<PartitionAssignment>>,
+}
+
+impl ClientContext for Context {}
+
+impl ConsumerContext for Context {
+    fn pre_rebalance(&self, rebalance: &Rebalance) {
+        info!("Pre rebalance {:?}", rebalance);
+    }
+
+    fn post_rebalance(&self, rebalance: &Rebalance) {
+        info!("Post rebalance {:?}", rebalance);
+
+        match rebalance {
+            Rebalance::Assign(tpl) => {
+                info!("Received new partition assignment list");
+                let partitions = partition_vec_from_topic_partition_list(tpl);
+                let partitions = Arc::new(partitions);
+                on_rebalance_assign(self.partition_assignment.clone(), partitions.clone());
+            }
+            Rebalance::Revoke => {
+                info!("Partition assignments revoked");
+                on_rebalance_revoke(self.partition_assignment.clone());
+            }
+            Rebalance::Error(e) => {
+                warn!(
+                    "Unexpected Kafka error in post_rebalance invocation {:?}",
+                    e
+                );
+            }
+        }
+    }
+}
+
+fn partition_vec_from_topic_partition_list(
+    topic_partition_list: &TopicPartitionList,
+) -> Vec<DataTypePartition> {
+    topic_partition_list
+        .to_topic_map()
+        .iter()
+        .map(|((_, p), _)| *p)
+        .collect()
+}
+
+fn on_rebalance_assign(
+    partition_assignment: Arc<Mutex<PartitionAssignment>>,
+    partitions: Arc<Vec<DataTypePartition>>,
+) {
+    let _fut = tokio::spawn(async move {
+        let mut partition_assignment = partition_assignment.lock().await;
+        partition_assignment.update_assignment(&partitions);
+    });
+}
+
+fn on_rebalance_revoke(partition_assignment: Arc<Mutex<PartitionAssignment>>) {
+    let _fut = tokio::spawn(async move {
+        let mut partition_assignment = partition_assignment.lock().await;
+        partition_assignment.clear();
+    });
+}
+
+impl Context {
+    pub fn new(partition_assignment: Arc<Mutex<PartitionAssignment>>) -> Self {
+        Self {
+            partition_assignment,
+        }
+    }
+}
+
+pub struct PartitionAssignment {
+    assignment: HashMap<DataTypePartition, Option<DataTypeOffset>>,
```

Comment:
There's a *lot* of work to manage the `PartitionAssignment` in the calling code in a `Mutex` to ostensibly make access safe. I'm wondering if adopting something like [dashmap](https://crates.io/crates/dashmap) would simplify the caller and implemented code here?

I believe the potential added benefit of bringing Dashmap into `PartitionAssignment` is that I _think_ that kafka-delta-ingest would have better performance since all partition update assignments are currently locking on this single mutex.

Consider this with a grain of salt, I'm not a TLA+ robot like @houqp so I don't yet have a complete model of how the concurrent flow works here.

Reply:
Interesting. Not familiar with dashmap. I will have a look. One thing to keep in mind is that the MutexGuard returned upon locking has to be a tokio mutex - cant be a std::sync::Mutex.

Reply:
Also - I'm personally not worried about _performance_ implications related to the `Arc<Mutex<PartititonAssignment>>` because these are pretty limited. We don't have to lock on every message. We only lock in a few scenarios:

1. init
2. rebalance
3. first run_loop iteration after rebalance

Having said that I am still concerned and highly interested in making the code around this mutex more readable and maintainable.

Reply:
@rtyler - after my first look at `dashmap` - i would like to try this out, but I'd like to hold exploration post-merge before diving too deep.